src/modules/uavcan/uavcan_configuration.cpp

```diff
@@ -0,0 +1,473 @@
+    client.setCallback([&](const uavcan::ServiceCallResult<T>& result)
+	{
+		success = result.isSuccessful();
+		response = result.response;
+	});
+ 
+    const int call_res = client.call(remote_node_id, request);
+    if (call_res >= 0)
+    {
+        while (client.isPending())
+        {
+            const int spin_res = node.spin(uavcan::MonotonicDuration::fromMSec(2));
+            if (spin_res < 0)
+            {
+                return {spin_res, response};
+            }
+        }
+        return {success ? 0 : -uavcan::ErrFailure, response};
+    }
+    return {call_res, response};
+}
+
+UavcanRemoteConfiguration::UavcanRemoteConfiguration(uavcan::INode &node) :
+	_mavlink_fd(-1),
+	_node(node),
+	_uavcan_pub_enumeration_cmd(node)
+{
+}
+
+UavcanRemoteConfiguration::~UavcanRemoteConfiguration()
+{
+}
+
+int UavcanRemoteConfiguration::init()
+{
+	_mavlink_fd = ::open(MAVLINK_LOG_DEVICE, 0);
+	return 0;
+}
+
+
+void UavcanRemoteConfiguration::uavcan_configuration(float val_1, float val_2, float val_3, float val_4, float val_5, float val_6, float val_7)
+{
+	switch ((int)val_1) {
+		case UAVCAN_CMD_PARAM_SAVE:
+			cmd_param_save((int)val_2, 0);
+			break;
+
+		case UAVCAN_CMD_PARAM_GET_ALL:
+			cmd_param_get_all((int)val_2);
+			break;
+
+		case UAVCAN_CMD_PARAM_GET:
+			cmd_param_get((int)val_2, (int)val_3);
+			break;
+
+		case UAVCAN_CMD_PARAM_SET:
+			cmd_set_param_set((int)val_2, (int)val_3, (float)val_4);
+			break;
+
+		case UAVCAN_CMD_DISCOVER_FOR_NODES:
+			cmd_discover_for_nodes();
+			break;
+
+		case UAVCAN_CMD_NODE_GET_INFO:
+			cmd_node_get_info((int)val_2);
+			break;
+
+		case UAVCAN_CMD_NODE_FACTORY_RESET:
+			cmd_param_save((int)val_2, 1);
+			break;
+
+		case UAVCAN_CMD_NODE_FACTORY_RESET_ALL_NODES:
+			mavlink_and_console_log_info(_mavlink_fd, "UAVCANPARAM: ERROR unsupported Command");
+			break;
+
+		case UAVCAN_CMD_NODE_RESTART:
+			cmd_node_restart((int)val_2);
+			break;
+
+		case UAVCAN_CMD_ENUMERATE:
+			cmd_enumerate((int)val_2, (int)val_3);
+			break;
+
+		default:
+			mavlink_and_console_log_info(_mavlink_fd, "UAVCANPARAM: ERROR unsupported Command");
+			break;
+	}
+}
+
+
+void UavcanRemoteConfiguration::cmd_param_save(int targetnode, int factoryreset)
+{
+	uavcan::protocol::param::SaveErase::Request msg;
+
+	msg.opcode = factoryreset;
+	auto res = performBlockingServiceCall<uavcan::protocol::param::SaveErase>(_node, targetnode, msg);
+
+	if (res.first < 0) {
+		// error
+		mavlink_and_console_log_info(_mavlink_fd, "UAVCANPARAM: ERROR Node:%d", targetnode);
+	} else {
+		if (factoryreset == 0) {
+			mavlink_and_console_log_info(_mavlink_fd, "UAVCANPARAM: Param-saved Node:%d", targetnode);
+		} else {
+			mavlink_and_console_log_info(_mavlink_fd, "UAVCANPARAM: Factory-reset Node:%d", targetnode);
+		}
+	}
+}
+
+
+void UavcanRemoteConfiguration::cmd_param_get_all(int targetnode)
+{
+	std::vector<uavcan::protocol::param::GetSet::Response> remote_params;
```

Comment:
Sorry for being late to join the discussion folks. `std::vector` is not totally ok, because:

- It may lead to heap fragmentation.
- Exceptions are disabled on the platform, so if the vector runs out of memory it'll likely abort the application.

So STL containers should not be used at all. A possible workaround is to send parameters via MAVLink immediately as they arrive from the remote node, rather than buffering them.